### DISCUSSION TITLE
Installing cfclient on Ubuntu 16.04

### DISCUSSION THREAD

**Discussion by TagirMuslimov** [bitcraze/discussions#398] [Q&A] [answered]
Hi all. 
Could you please tell me how to install older version of cflient on Ubuntu 16.04 without errors?
I need it for [sim_cf](https://github.com/wuwushrek/sim_cf) simulation.
For sim_cf simulation in HITL mode I need 

> cd ~/catkin_ws/src/sim_cf/crazyflie-firmware
> make PLATFORM=hitl
> Start crazyflie in bootloader mode then 
> make cload

Maybe the cfclient is not needed to do this?
This is what I got when I tried to install a modern сfclient.

> tagir@tagir-Vostro-5468:~/crazyflie-clients-python$ pip3 install --user -e .Obtaining file:///home/tagir/crazyflie-clients-python
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/home/tagir/crazyflie-clients-python/setup.py", line 51, in <module>
>         raise "must use python 3.7 or greater"
>     TypeError: exceptions must derive from BaseException
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /home/tagir/crazyflie-clients-python/
> You are using pip version 8.1.1, however version 22.3.1 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.

From this error I understand that I need a newer version of Python or an older version of cfclient.
However, on Ubuntu 16 I have Python version 3.5 installed. And I tried to do it with the new version of Python 3.7 in a virtual machine, but it didn't work.
How do I install an older version of cfclient?

**Comment by victorhook** [comment]
Hey!

Have you tried installing cfclient with from pypi? Then you can specify exactly which version you want to install, eg: `pip install cfclient==2022.3`
You can find all versions here: https://pypi.org/project/cfclient/#history

**Reply by TagirMuslimov** [reply]
I tried pip and it says Python version is wrong
> tagir@tagir-Vostro-5468:~$ pip install cfclient==2017.6
> Collecting cfclient==2017.6
>   Downloading https://files.pythonhosted.org/packages/d4/7c/d87beb5612002ab3c9eccde09144e6368c123bbdfd3a24a4c91cd937d0a1/cfclient-2017.6.tar.gz (108kB)
>     100% |████████████████████████████████| 112kB 866kB/s 
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-QQLUp5/cfclient/setup.py", line 18, in <module>
>         raise "must use python 3.4 or greater"
>     TypeError: exceptions must be old-style classes or derived from BaseException, not str
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-QQLUp5/cfclient/
> You are using pip version 8.1.1, however version 22.3.1 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.


I tried pip3 and got some strange error

> tagir@tagir-Vostro-5468:~$ pip3 install cfclient==2017.6
> Collecting cfclient==2017.6
>   Downloading https://files.pythonhosted.org/packages/84/e4/4a68981d5e656132c62b9b5c3186ba61bc9eb8249a0cf2a7e743c3b6eac4/cfclient-2017.6-py3-none-any.whl (173kB)
>     100% |████████████████████████████████| 174kB 1.2MB/s 
> Collecting pyzmq (from cfclient==2017.6)
>   Downloading https://files.pythonhosted.org/packages/46/0d/b06cf99a64d4187632f4ac9ddf6be99cd35de06fe72d75140496a8e0eef5/pyzmq-24.0.1.tar.gz (1.2MB)
>     100% |████████████████████████████████| 1.2MB 1.0MB/s 
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-kkt4n4xp/pyzmq/setup.py", line 165
>         print(f"DEBUG EXT_SUFFIX {suffix!r} {ext_suffix!r}")
>                                                           ^
>     SyntaxError: invalid syntax
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-kkt4n4xp/pyzmq/
> You are using pip version 8.1.1, however version 22.3.1 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.

**Reply by victorhook** [reply]
Hmm, I would suggest using virtual environments when using python, which is extremely useful when you want specific versions of packages.
Maybe you can try:
1. Create a new directory.
2. Enter the directory.
3. `python3 -m venv env`
4. `source env/bin/acitvate`
5. `pip install cfclient==2017.6`

This works on my machine, but I'm running Ubuntu 22.04

**Reply by TagirMuslimov** [reply]
Unfortunately, it didn't work.
Here's what I got
> (env) tagir@tagir-Vostro-5468:~/for_cfclient$ python3 -m venv env
> (env) tagir@tagir-Vostro-5468:~/for_cfclient$ source env/bin/activate
> (env) tagir@tagir-Vostro-5468:~/for_cfclient$ pip install cfclient==2017.6
> Collecting cfclient==2017.6
>   Using cached https://files.pythonhosted.org/packages/84/e4/4a68981d5e656132c62b9b5c3186ba61bc9eb8249a0cf2a7e743c3b6eac4/cfclient-2017.6-py3-none-any.whl
> Collecting pyqtgraph>=0.10 (from cfclient==2017.6)
>   Using cached https://files.pythonhosted.org/packages/a2/88/8f2e85c0e171b7b08357f61551614ff39c4cab644a242bc80907e2275e47/pyqtgraph-0.13.1-py3-none-any.whl
> Collecting appdirs==1.4.0 (from cfclient==2017.6)
>   Using cached https://files.pythonhosted.org/packages/7b/8b/eebc6e2002a1e0383f1c7108d0111d4d33ea93bf417d7e19e43ec9b87b2b/appdirs-1.4.0-py2.py3-none-any.whl
> Collecting pyzmq (from cfclient==2017.6)
>   Using cached https://files.pythonhosted.org/packages/46/0d/b06cf99a64d4187632f4ac9ddf6be99cd35de06fe72d75140496a8e0eef5/pyzmq-24.0.1.tar.gz
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-gudc0gx_/pyzmq/setup.py", line 165
>         print(f"DEBUG EXT_SUFFIX {suffix!r} {ext_suffix!r}")
>                                                           ^
>     SyntaxError: invalid syntax
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-gudc0gx_/pyzmq/

**Reply by victorhook** [reply]
Hmm, that error looks a bit weird to me.
Are you running it with python 3?
Perhaps you can try running `pip install` with `--no-cache-dir` flag?

**Reply by TagirMuslimov** [reply]
> Perhaps you can try running pip install with --no-cache-dir flag? 

The same error

Maybe there's something wrong with the Python version? But I ran the command and it seems to be the right version.
> (env) tagir@tagir-Vostro-5468:~/for_cfclient$ python --version
> Python 3.5.2

I tried
> (env) tagir@tagir-Vostro-5468:~/for_cfclient$ python3 -m pip install cfclient==2017.6

The same result
> ...
>   Using cached https://files.pythonhosted.org/packages/46/0d/b06cf99a64d4187632f4ac9ddf6be99cd35de06fe72d75140496a8e0eef5/pyzmq-24.0.1.tar.gz
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-7a664z8p/pyzmq/setup.py", line 165
>         print(f"DEBUG EXT_SUFFIX {suffix!r} {ext_suffix!r}")
>                                                           ^
>     SyntaxError: invalid syntax
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-7a664z8p/pyzmq/

**Comment by TagirMuslimov** [comment, marked as answer]
@victorhook Thank you so much for trying to help me. After much struggling, I was able to install the client and solve the problem. I will write here so that I do not forget in the future:

> tagir@tagir-Vostro-5468:~$ curl -fsSL https://bootstrap.pypa.io/pip/3.5/get-pip.py | python3.5

What I got

> DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.
> Defaulting to user installation because normal site-packages is not writeable
> Collecting pip<21.0
>   Downloading pip-20.3.4-py2.py3-none-any.whl (1.5 MB)
>      |████████████████████████████████| 1.5 MB 1.1 MB/s 
> Installing collected packages: pip
> Successfully installed pip-20.3.4

> tagir@tagir-Vostro-5468:~$ pip install cfclient==2017.6

And this is not the end. The client will not start if you do not make

> tagir@tagir-Vostro-5468:~$ sudo apt-get install python3-pyqt5
